include/zephyr/drivers/sensor_v2.h

```diff
@@ -0,0 +1,989 @@
+	size_t offset;
+};
+
+/**
+ * @brief A helper macro for assigning an iterator
+ *
+ * @param raw_data Pointer to the sensor_raw_data containing the data to iterate over
+ */
+#define SENSOR_FIFO_ITERATOR_OF(raw_data)                                                          \
+	{                                                                                          \
+		.data = raw_data,                                                                  \
+		.offset = 0,                                                                       \
+	}
+
+/**
+ * @typedef sensor_fifo_iterator_next_t
+ * @brief FIFO iterator API function to move the iterator to the next value
+ *
+ * @see sensor_fifo_iterator_api.next for argument description
+ */
+typedef int (*sensor_fifo_iterator_next_t)(struct sensor_fifo_iterator *iterator);
+
+/**
+ * @typedef sensor_fifo_iterator_get_sensor_type_t
+ * @brief FIFO iterator API function to get the sensor type of the current iterator position
+ *
+ * @see sensor_fifo_iterator_api.get_sensor_type for argument description
+ */
+typedef int (*sensor_fifo_iterator_get_sensor_type_t)(const struct sensor_fifo_iterator *iterator,
+						      uint32_t *sensor_type);
+
+/**
+ * @typedef sensor_fifo_iterator_read_t
+ * @brief FIFO iterator API function to read the current iterator position into SI units
+ *
+ * @see sensor_fifo_iterator_api.read for argument description
+ */
+typedef int (*sensor_fifo_iterator_read_t)(const struct sensor_fifo_iterator *iterator,
+					   void *out);
+
+/**
+ * @brief The FIFO iterator API as described by the individual sensor
+ */
+struct sensor_fifo_iterator_api {
+	/**
+	 * @brief Move the iterator to the next position if possible
+	 *
+	 * @param iterator The iterator to move
+	 * @return 0 if the iterator was moved
+	 * @return < 0 on failure
+	 */
+	sensor_fifo_iterator_next_t next;
+
+	/**
+	 * @brief Get the sensor type of the current iterator position
+	 *
+	 * @param iterator The iterator to query
+	 * @param sensor_type Pointer to the sensor type that will be written on success
+	 * @return 0 on success
+	 * @return < 0 on failure
+	 */
+	sensor_fifo_iterator_get_sensor_type_t get_sensor_type;
+
+	/**
+	 * @brief Read the current iterator position as SI units
+	 *
+	 * It is the responsibility of the caller to make sure that the out argument type matches up
+	 * with the current iterator's sensor type. This means that if get_sensor_type returns
+	 * SENSOR_TYPE_ACCELEROMETER, then out must be of type sensor_three_axis_data.
+	 *
+	 * @param iterator The iterator to query
+	 * @param out The output data structure to write to
+	 * @return 0 on success
+	 * @return < 0 on failure
+	 */
+	sensor_fifo_iterator_read_t read;
+};
+
+/**
+ * @typedef sensor_read_data_t
+ * @brief Sensor API function for reading a single data point
+ *
+ * @see sensor_read_data() for argument description
+ */
+typedef int (*sensor_read_data_t)(const struct device *sensor, uint32_t sensor_type, void *data);
```

Comment:
won't it have more sense to pass data type instead of sensor type?

Reply:
users should already know the sensor type as they open the device
this would be practical only in case of one callback used for multiple sensors but still it will be better to use pass information about data type as it is more generic (not only to easier handle new sensor types but also for cases when sensor allow to switch data type).

it might also be better to simply leave this for user to check using a separate call (e.g. sensor_get_type) which would be useful if type of the sensor is known (expect that this would be most cases).

Reply:
Not necessarily. The BMI160 supports 4 different types: accel, gyro, accel_temp, and gyro_temp. For these accel/gyro share the same data type, and both temperature readings use the same data type. So we need to tell the driver what we're trying to read.

Reply:
True - there was a discussion about that in the comments, that I read later.

In such case I recommend to not use explicit names (enums, defines here) but pass instance id in the argument that api user can use to get more information about the specific instance. This way the API would be easily expandable to other types of sensors. Also you can get more information about this specific instance easy if that would be needed.

e.g.
```
int (*sensor_read_data_t)(const struct device *sensor, uint32_t sensor_instance, void *data);

char *sensor_get_instance_units(const struct device *sensor, uint32_t sensor_instance); // returns "m/s^2" "g" "g" kg", etc
char *sensor_get_instance_type(const struct device *sensor, uint32_t sensor_instance); // returns "accelerometer" "magnetometer", etc
```

The units/types could be returned as integers (or enums) but then you would have to define everything. For smaller projects it may not be practical (I doubt anybody will check if for the single sensor that is attached) and for larger and future it might be limiting.

Anyway this is a general idea.